pkg/informer/informer.go

```diff
@@ -24,92 +24,182 @@ import (
+)
 
 // DynamicDiscoverySharedInformerFactory is a SharedInformerFactory that
 // dynamically discovers new types and begins informing on them.
 type DynamicDiscoverySharedInformerFactory struct {
-	workspaceLister tenancylisters.ClusterWorkspaceLister
-	disco           clusterDiscovery
-	dynamicClient   dynamic.Interface
-	filterFunc      func(interface{}) bool
-	pollInterval    time.Duration
-	indexers        cache.Indexers
+	dynamicClient dynamic.Interface
+	filterFunc    func(interface{}) bool
+	indexers      cache.Indexers
+	crdIndexer    cache.Indexer
 
 	// handlersLock protects multiple writers racing to update handlers.
 	handlersLock sync.Mutex
 	handlers     atomic.Value
 
-	mu               sync.RWMutex
+	updateCh chan struct{}
+
+	informersLock    sync.RWMutex
 	informers        map[schema.GroupVersionResource]informers.GenericInformer
 	startedInformers map[schema.GroupVersionResource]bool
 	informerStops    map[schema.GroupVersionResource]chan struct{}
 	terminating      bool
+	terminatedCh     chan struct{}
+}
+
+// NewDynamicDiscoverySharedInformerFactory returns a factory for shared
+// informers that discovers new types and informs on updates to resources of
+// those types.
+func NewDynamicDiscoverySharedInformerFactory(
+	dynClient dynamic.Interface,
+	filterFunc func(obj interface{}) bool,
+	crdInformer apiextensionsinformers.CustomResourceDefinitionInformer,
+	indexers cache.Indexers,
+) (*DynamicDiscoverySharedInformerFactory, error) {
+	f := &DynamicDiscoverySharedInformerFactory{
+		dynamicClient: dynClient,
+		filterFunc:    filterFunc,
+		indexers:      indexers,
+		crdIndexer:    crdInformer.Informer().GetIndexer(),
+
+		// Use a buffered channel of size 1 to allow enqueuing 1 update notification
+		updateCh: make(chan struct{}, 1),
+
+		informers:        make(map[schema.GroupVersionResource]informers.GenericInformer),
+		startedInformers: make(map[schema.GroupVersionResource]bool),
+		informerStops:    make(map[schema.GroupVersionResource]chan struct{}),
+		terminatedCh:     make(chan struct{}),
+	}
+
+	f.handlers.Store([]GVREventHandler{})
+
+	// Add an index function that indexes a CRD by its group/firstServedVersion/resource. We only need the first
+	// served version because this shared informer factory is expected to be using a wildcard client for partial
+	// metadata only. In this instance, version does not matter, because a wildcard partial metadata list request
+	if err := crdInformer.Informer().AddIndexers(cache.Indexers{
+		groupFirstFoundVersionResourceIndex: func(obj interface{}) ([]string, error) {
+			crd, ok := obj.(*apiextensionsv1.CustomResourceDefinition)
+			if !ok {
+				return nil, fmt.Errorf("%T is not a CustomResourceDefinition", obj)
+			}
+
+			firstServedVersion := ""
+			for _, version := range crd.Spec.Versions {
+				if !version.Served {
+					continue
+				}
+				firstServedVersion = version.Name
+				break
+			}
+
+			if firstServedVersion == "" {
```

Comment:
When is this true?

Reply:
Probably never, but this was a "just in case"